### DISCUSSION TITLE
Format of design specs

### DISCUSSION THREAD

**Discussion by kgodey** [mathesar-foundation/mathesar#279] [Maintainer discussions]
Now that we've gone through the design spec process for a few features, I wanted to start a discussion on the format that we're using for the specs.

Questions:
- @ghislaineguerin, 
  - how easy is it for you to produce the specs?
  - what would make it easier?
  - Are there areas that you find useful or not useful?
  - are there areas that take a lot of work but change quickly that you'd like to omit?
- @pavish, 
  - does the current format work well for implementing the spec? 
  - what additional details would you like to see?
  - are there any details you consider superfluous?
- @mathemancer @pavish @eito-fis,
  - does the current format work well for reviewing the spec? 
  - are there additional details you'd like to see?
  - are there details you consider superfluous?
  - how does the spec format work for you on re-reviews?
  - are there changes you'd like to see to the re-review process?

The issue that prompted this discussion is that it's hard to do re-reviews with the current format, since the screenshots and explanations end up being out of sync with the prototype, and I can't comment on individual sections. I'll answer the questions above in more detail in a comment.

I don't want to jump to proposing solutions yet, so let's stick to gathering information first.

**Comment by ghislaineguerin** [comment]
### How easy is it for you to produce the specs?
I'm not sure if it's a matter of how easy it is, rather than It's harder to design if I don't produce the specs. Usually, as I'm creating the different designs on Figma, I add them to the document, and it helps me notice anything that might be missing or scenarios I hadn't considered.

The difficulty, I think, comes with the scope of each design task and how many problems we are expecting to solve in a single iteration. There are multiple constraints on different design scenarios (i.e., can we partially save a record with an error?), so if we might get into a case where we don't know enough to decide, I break up the design and solve the problem most immediate problems. So let's say, for example, in the case of 'Adding Records,' the goal is to define how a user might add a new record to a table. However, we also need to figure out how this interacts with filters, how errors are handled, how auto-save should work, and more. Adding all of that into the specs would make the document unmanageable. For example, if we decide how to display error messages within the table, but we don't try that solution when a user adds 100 rows with errors by pasting them into the table, we might have to redesign it down the road.

That being said, this is not always a matter of the task being too difficult, but the time it would take to complete the design for most scenarios. Once we have a better idea of the product and how the different parts work together, it can be faster.

### What would make it easier?
To reduce the scope of the solutions until we have a shorter cycle from definition to test. For me, that would be a version of the product in which the team can test design solutions in the context of the problem we are trying to solve.

### Are there areas that you find useful or not useful?
I think discussing the different interactions introduced is helpful as I can describe more scenarios or interdependences between new components and existing ones. I'd go deeper into this if we reduced the scope.
Also, consider that I write the specs for myself too, so details that might seem not relevant result from questions or experiences that haven't been fully explored but might in the future, and it's a good way for others to understand my reasoning behind certain design decisions.

### Are there areas that take a lot of work but change quickly that you'd like to omit?
Over-defining the UI or UX Copy at this point. For now, I want to have a high level of understanding of the design in terms of which components are required for a user to input data, how users can save data, etc., and other things that prevent users from completing a task within Mathesar.
Once we have that, I can more easily dive into the areas that require more fine-grained interactions.

**Comment by pavish** [comment]
### Does the current format work well for implementing the spec?

The current format works good for me.

However, when implementing the spec, I sometimes notice that there are some details that we have glossed over while discussing the design. For eg., with the pagination. I believe there are a lot of factors to it, the major one being that some interactions aren't usually visible until they are being implemented, the secondary one being the product definition itself. I believe we need to iron out this problem.

Also, when we work on the design we tend to see it on a higher vantage point. But unless it is detailed down to a particular level, we cannot proceed with implementation, or if we do, we might end up with a lot of rework later on.

For eg.,

1. The sidebar has a tree with schema and tables. If the table count exceeds 100, the sidebar might appear cluttered. If we decide to change the design later on (where we do not fetch all table details in a single api call), the work on the UI would need a bit of refactoring which will take more effort than implementing it new.
2. The UI currently assumes that only one DB can be chosen at a time. Having said that, this implies that the tables that are open belong to a single db. If we have to support tables from multiple dbs to be open at the same time, the url structure would have to change, the internal store structure would need changes. Since this is more of a base functionality, the more code we build over it, the more time it will take to refactor.

I would like to avoid this, or keep it to a minimum.

### What additional details would you like to see?

I believe the previous answer also answers this question.

### Are there any details you consider superfluous?

No, I do not. I think more details are good.

### Does the current format work well for reviewing the spec?

Yes. Taking it out onto GH discussions works good.

### How does the spec format work for you on re-reviews?

I think we have to improve re-reviews. Sometimes it is not clear which part of the document is updated based on the review, and which isn't.

### Are there changes you'd like to see to the re-review process?

Yeah, I think after every review process, it would be best if we added a comment on the GH discussion, with the review points taken into account, and what are the changes that are to be done based on those points. And also if a review point is not considered, the reasons for not taking them. It is easier for me to refer the GH discussion, rather than on hackMD.

This might improve our re-review process.

**Reply by ghislaineguerin** [reply]
I agree with your points on re-reviews and I like your idea of adding the conclusions to the GH discussion + document.

**Comment by mathemancer** [comment]
> does the current format work well for reviewing the spec?

I find it a little difficult to sync the discussion flow between hackMD and github discussions.  I'm not sure there's a better way, though, unless we can get all we want on GH (can Figma widgets embed here?).

> are there additional details you'd like to see?

None that I can think of.

> are there details you consider superfluous?

No.

> how does the spec format work for you on re-reviews?

I think it's similar to what @kgodey mentioned; it's tricky to keep up with what was changed, what wasn't, and trying to remember what I commented before.

> are there changes you'd like to see to the re-review process?

I'd like:
- all discussion in one place.
- PRs, with changes highlighted between rounds (as they would be with Github PRs)

I realize this might not be possible to achieve, but I thought I'd write out my ideal scenario.

**Comment by kgodey** [comment]
My answers:

> does the current format work well for reviewing the spec?

I find reviewing on GitHub Discussions easier than reviewing on HackMD because it's easier to follow the discussion. Inline comments are a benefit of HackMD, but their usefulness is limited after the first review, and I'd be willing to give them up entirely.

> are there additional details you'd like to see?

None that I can think of.

> are there details you consider superfluous?

I'm not sure that we need screenshots in the specs when we have the prototype. Things change quickly on review, and the screenshots get out of date. Keeping the spec as text and referencing the prototype via links seems like it would work better.

> how does the spec format work for you on re-reviews?

Not very well, it's hard to see what's changed.

> are there changes you'd like to see to the re-review process?

I agree that being able to see diffs highlighted would be useful.

**Comment by kgodey** [comment]
Based on the information gathered so far, I'd like to propose the following changes to the review process:
- Specs should be written directly on the wiki, rather than HackMD, and all the discussion around them should be on GitHub Discussions. This has the following advantages:
  1. centralizes the conversation in one place
  2. gives us good diffs for changes that we can link to (through the Wiki.js version history page) 
  3. avoids the step of having to push HackMD specs to the wiki
  4. makes our process even more transparent.
- Specs should not include screenshots, only text and links to the prototype. The text should match the updates in the prototype before re-review is requested so that the entire document fits together well. This should:
   1. Make re-review easier for the reviewers because the prototype matches the text descriptions.
   2. Make updating the document easier for the author, because new screenshots are not constantly necessary
   3. Make diffs easier to read.
- When requesting re-review @ghislaineguerin should add a comment on the GitHub discussion summarizing the review points from the last round of feedback, and explaining what changes have been made and not made (if a review point resulted in changes not being made, the reasoning should be explained).

I think we should follow the new format starting with the next new spec. Thoughts/feedback?

**Reply by mathemancer** [reply]
I think this makes sense, and I'd like to try it.  It would be nice to "clump" the edits together so we can see things at the "finished revision" level of granularity rather than commit-by-commit.  Maybe we could record the commit hash at the initial review, then also the one at the next review, so that those with the repo pulled locally can `git diff` the relevant  commits to see the whole revision's changes at once.

**Reply by kgodey** [reply]
@ghislaineguerin @pavish @mathemancer I added a third point above, which was @pavish's suggestion for improving re-review.

@mathemancer Wiki.js has an inbuilt version history page where we can select the versions to compare and see the diff, we don't need to use `git`. e.g. see https://wiki.mathesar.org/h/en/home. Of course, you can also use `git` if you prefer.
